File: 1407-group-the-people-given-the-group-size-they-belong-to/1407-group-the-people-given-the-group-size-they-belong-to.java
class Solution {
    public List<List<Integer>> groupThePeople(int[] arr) {
        Map<Integer,List<Integer>> mp=new HashMap<>();
        for(int i=0;i<arr.length;++i){
            List<Integer> currList=mp.getOrDefault(arr[i],new ArrayList<>());
            currList.add(i);
            mp.put(arr[i],currList);
        }
        List<List<Integer>> ans=new ArrayList<>();
        for(Map.Entry<Integer,List<Integer>> entry:mp.entrySet()){
            int requiredSize=entry.getKey();
            List<Integer> val=entry.getValue();
            List<Integer> curr=new ArrayList<>();
            for(int i=0;i<val.size();++i){
                curr.add(val.get(i));
                if(curr.size()==requiredSize){
                    ans.add(curr);
                    curr=new ArrayList<>();
                }
            }
        }
        return ans;
    }
}

File: 1407-group-the-people-given-the-group-size-they-belong-to/1407-group-the-people-given-the-group-size-they-belong-to.cpp
class Solution {
public:
    vector<vector<int>> groupThePeople(vector<int>& grp) {
        vector<vector<int>> arr(grp.size());
        for(int i=0;i<grp.size();++i){
            arr[i]={grp[i],i};
        }
        sort(arr.begin(),arr.end());
        vector<vector<int>> ans;
        int i=0;
        while(i<arr.size()){
            int j=i;
            while(j<arr.size()&&j-i+1<=arr[i][0]&&arr[j][0]==arr[i][0])++j;
            if(j-i<arr[i][0])return {};
            vector<int> curr;
            for(int k=i;k<j;++k)curr.push_back(arr[k][1]);
            ans.push_back(curr);
            i=j;
        }
        return ans;
    }
};